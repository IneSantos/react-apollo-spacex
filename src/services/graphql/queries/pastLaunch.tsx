import { gql } from "@apollo/client";

export const GET_PAST_LAUNCH = gql`
  query getLaunch($launchID: ID!) {
  launch(id: $launchID) {
    details
    id
    links {
      article_link
      flickr_images
      video_link
    }
    mission_name
    ships {
      id
      image
    }
  }
}
`