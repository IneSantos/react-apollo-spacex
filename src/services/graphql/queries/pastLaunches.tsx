import { gql } from "@apollo/client";

export const GET_PAST_LAUNCHES = gql`
  query getPastLaunches {
  launchesPast(limit: 10) {
    id
    mission_name
    launch_date_local
    details
  }
}
`