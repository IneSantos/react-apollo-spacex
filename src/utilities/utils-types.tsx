
export interface Launches {
    id: string,
    mission_name: string,
    launch_date_local: string,
    details: string | null
}

interface Ship {
    id: string,
    image: string
}

export interface Launch {
    id: string,
    mission_name: string,
    details: string,
    links: {
        article_link: string | null,
        flickr_images: string[],
        video_link: string | null
    },
    ships: Ship[]
}