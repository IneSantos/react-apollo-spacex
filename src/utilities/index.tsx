import { Mission } from "../components/atomic-ui/organisms/mission-small-card/mission-type";
import { Mission as MissionCard } from "../components/atomic-ui/organisms/mission-card/mission-type";
import { Launch, Launches } from "./utils-types";

export const getPastMissions = (launches: Launches[]): Mission[] => {
    return launches.map((launch) => ({
        id: launch.id,
        name: launch.mission_name,
        date: launch.launch_date_local,
        details: launch.details || ""
    }))
}


export const getMission = (launch: Launch): MissionCard => {
    const img = launch.links.flickr_images.length > 0 ? launch.links.flickr_images[0] : (launch.ships.length > 0 ? launch.ships[0].image : "");

    return {
        name: launch.mission_name,
        link: launch.links.article_link || launch.links.video_link || "",
        details: launch.details || "",
        img
    }
}

export const formattedDate = (date: string) => {
    const d = new Date(date);
    const day = d.getDate();
    const year = d.getFullYear();

    const months = [
        'January',
        'February',
        'March',
        'April',
        'May',
        'June',
        'July',
        'August',
        'September',
        'October',
        'November',
        'December'
    ];
    const monthName = months[d.getMonth()];

    return `${monthName} ${day}, ${year}`
}