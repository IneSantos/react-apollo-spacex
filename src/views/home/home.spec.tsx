import React from "react";
import { act, render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { GraphQLError } from "graphql";
import { GET_PAST_LAUNCHES } from "../../services/graphql/queries/pastLaunches";
import { MockedProvider } from '@apollo/client/testing';
import Home from "./home";
import { Launches } from "../../utilities/utils-types";

describe("Home view tests", () => {
    test(`Home component render with loader`, () => {
        const { container } = render(
        <MockedProvider mocks={[]}>
            <Home/>
        </MockedProvider>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    test(`Home component render with success`, async () => {
        const launches: Launches[] = [
            {
                id: "0",
                mission_name: "This is a mission name",
                details: "This is a mission detail",
                launch_date_local: "2020-10-24T11:31:00-04:00"
            }
        ]

        const mocks = [{
            request: {
                query: GET_PAST_LAUNCHES
            },
            result: {
                data: { launchesPast: launches}
            },
        }];

        const { container } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <Router>
                <Home/>
            </Router>
        </MockedProvider>
        );

        await act(() => new Promise((resolve) => setTimeout(resolve, 0)));
        expect(container.firstChild).toMatchSnapshot()
    });

    test(`Home component render with error msg`, async () => {
        const mocks = [{
            request: {
                query: GET_PAST_LAUNCHES
            },
            result: {
                errors: [new GraphQLError('Error!')]
            },
        }];

        const { container } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <Home/>
        </MockedProvider>
        );
    
        await act(() => new Promise((resolve) => setTimeout(resolve, 0)));
        expect(container.firstChild).toMatchSnapshot();
    });
});
