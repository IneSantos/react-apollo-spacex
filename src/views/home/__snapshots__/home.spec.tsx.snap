// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home view tests Home component render with error msg 1`] = `
<p>
  Error :(
</p>
`;

exports[`Home view tests Home component render with loader 1`] = `
<div
  class="loder"
>
  <div
    class="lds-ellipsis"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`Home view tests Home component render with success 1`] = `
<div
  class="spacex-missions"
>
  <header>
    <div
      class="banner"
    />
    <h1>
      Recent Launches
    </h1>
  </header>
  <section
    class="missions-body"
  >
    <div
      class="mission-list"
    >
      <a
        href="/mission/0"
        style="text-decoration: none;"
      >
        <div
          class="mission-wrapper"
        >
          <div
            class="card-header"
          >
            <h2
              class="title"
            >
              This is a mission name
            </h2>
          </div>
          <div
            class="list-card-body"
          >
            <div
              class="details"
            >
              This is a mission detail
            </div>
            <div
              class="launch-date"
            >
              October 24, 2020
            </div>
          </div>
        </div>
      </a>
    </div>
  </section>
</div>
`;
